File: frontend/src/realtime/events.ts
import type { Message } from '../types/chat'

export interface ConnectionReadyEvent {
  type: 'connection.ready'
  data: { user_id: string }
}

export interface MessageCreatedEvent {
  type: 'message.created'
  request_id: string | null
  data: Message
}

export interface ErrorEvent {
  type: 'error'
  request_id: string | null
  error: { code: string; message: string }
}

export type ServerEvent = ConnectionReadyEvent | MessageCreatedEvent | ErrorEvent

export interface MessageSendEvent {
  type: 'message.send'
  request_id: string
  conversation_id: string
  content: string
}

export function isServerEvent(value: unknown): value is ServerEvent {
  if (!value || typeof value !== 'object' || !('type' in value)) return false
  const event = value as Record<string, unknown>
  if (event.type === 'connection.ready') {
    const data = event.data as Record<string, unknown> | undefined
    return typeof data?.user_id === 'string'
  }
  if (event.type === 'message.created') {
    const data = event.data as Record<string, unknown> | undefined
    return typeof data?.id === 'string'
      && typeof data.conversation_id === 'string'
      && typeof data.content === 'string'
      && typeof data.cursor === 'string'
  }
  if (event.type === 'error') {
    const error = event.error as Record<string, unknown> | undefined
    return typeof error?.code === 'string' && typeof error.message === 'string'
  }
  return false
}

File: backend/app/websocket/events.py
from typing import Literal, Optional, Union
from uuid import UUID

from pydantic import BaseModel, Field, TypeAdapter, field_validator

from app.schemas.message import MessageResponse


class MessageSendEvent(BaseModel):
    type: Literal["message.send"]
    request_id: Optional[str] = Field(default=None, max_length=64)
    conversation_id: UUID
    content: str = Field(min_length=1, max_length=4000)

    @field_validator("content")
    @classmethod
    def reject_blank_content(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("Message cannot be blank")
        return value


ClientEvent = Union[MessageSendEvent]
client_event_adapter = TypeAdapter(ClientEvent)


class ConnectionReadyEvent(BaseModel):
    type: Literal["connection.ready"] = "connection.ready"
    data: dict[str, str]


class MessageCreatedEvent(BaseModel):
    type: Literal["message.created"] = "message.created"
    request_id: Optional[str] = None
    data: MessageResponse


class ErrorDetail(BaseModel):
    code: str
    message: str


class ErrorEvent(BaseModel):
    type: Literal["error"] = "error"
    request_id: Optional[str] = None
    error: ErrorDetail
